docs/proposals/20250722-openstackclusteridentity.md

```diff
@@ -0,0 +1,384 @@
+    - key: kubernetes.io/metadata.name
+      operator: In
+      values: [team-a, team-b]
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackClusterIdentity
+metadata:
+  name: development-openstack
+spec:
+  secretRef:
+    name: dev-openstack-credentials
+    namespace: capo-system
+  namespaceSelector:
+    matchLabels:
+      environment: development
+```
+
+### Current Secret-based Identity (unchanged)
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackCluster
+metadata:
+  name: cluster-a
+  namespace: team-a
+spec:
+  identityRef:
+    # No type field = defaults to Secret behavior
+    name: my-secret
+    cloudName: openstack
+```
+
+### New ClusterIdentity-based (new usage)
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackCluster
+metadata:
+  name: cluster-a
+  namespace: team-a
+spec:
+  identityRef:
+    type: ClusterIdentity
+    name: prod-openstack
+```
+
+### Explicit Secret Type (optional)
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OpenStackCluster
+metadata:
+  name: cluster-a
+  namespace: team-a
+spec:
+  identityRef:
+    type: Secret
+    name: my-secret
+    cloudName: openstack
+```
+
+## Implementation Notes
+
+**API Changes**: 
+- Extend `OpenStackIdentityReference` with `type` field with `Secret` and `ClusterIdentity` as the only supported values
+- Add CEL validation rules for type-specific field combinations
+- Update CRD generation for new field
+
+**RBAC Changes**:
+- Add namespace `get` permission to existing `config/rbac/role.yaml`
+- Add `openstackclusteridentities` resource permissions
+
+**Controller Changes**: 
+- Modify `pkg/scope/provider.go` for type-based credential resolution
+- Add namespace access validation logic in controllers
+- Add cluster identity lookup and permission checking
+- Implement comprehensive error handling with clear messages
+
+**Backward Compatibility**: 
+- Default `type` to `Secret` when not specified
+- CEL validation maintains existing field requirements for secret-based identities
+- Ensure all existing configurations continue working
+
+**Security Considerations**:
+- Validate namespace access on every cluster reconciliation
+- Log access attempts for audit purposes
+- Clear error messages for permission denials
+- Fail securely when cluster identity is not accessible
+
+**Broader Impact**: 
+This change automatically enables cluster identity support for:
+- `OpenStackCluster` resources
+- `OpenStackMachine` resources  
+- `OpenStackServer` resources
+
+## Conditions
+
+Following the existing CAPO pattern for simple resources (like OpenStackFloatingIPPool), the `OpenStackClusterIdentityStatus` uses a single condition:
+
+- **Ready**: Indicates whether the identity can be used for authentication. This includes secret accessibility, structure validation, and credential validation. Different reason codes provide specific failure details.
+
+### Example Condition States
```

Comment:
Everything below implies:

* The existence of a ClusterIdentity controller
* That resource controllers consuming a ClusterIdentity must:
  - Wait for their ClusterIdentity to have the Ready condition, and therefore
  - Watch the ClusterIdentity resource

Is that complexity required in the first instance? What if it was just a dumb data struct, and any errors were reported by the resource controllers as normal?

Reply:
Done. I have removed the conditions field for now. We can consider it later if required.